extensions/amp-subscriptions/0.1/platform-store.js

```diff
@@ -374,60 +392,83 @@ export class PlatformStore {
    * @private
    */
-  selectApplicablePlatform_() {
+  selectApplicablePlatform_(optionalFactor) {
     const localPlatform = this.getLocalPlatform();
     let localWeight = 0;
-    /** @type {!Array<!Object<!./subscription-platform.SubscriptionPlatform, number>>} */
-    const platformWeights = [];
 
     dev().assert(this.areAllPlatformsResolved_(),
         'All platforms are not resolved yet');
 
-    this.getAvailablePlatforms().forEach(platform => {
-      let weight = 0;
+    /// Subscriber wins immediatly.
+    const availablePlatforms = this.getAvailablePlatforms();
+    for (const platformIdx in availablePlatforms) {
+      const platform = availablePlatforms[platformIdx];
       const entitlement =
           this.getResolvedEntitlementFor(platform.getServiceId());
-
-      // Subscriber wins immediatly.
       if (entitlement.isSubscriber()) {
-        weight += 100000;
+        return Promise.resolve(platform);
       }
-
-      // Add the base score
-      weight += platform.getBaseScore();
-
-      // If supports the current viewer, gains weight 9
-      if (platform.supportsCurrentViewer()) {
-        weight += this.scoreConfig_['supportsViewer'];
-      }
-
-      platformWeights.push({
-        platform,
-        weight,
-      });
-      if (platform.getServiceId() === 'local') {
-        localWeight = weight;
-      }
-    });
-
-    platformWeights.sort(function(platform1, platform2) {
-      return platform2.weight - platform1.weight;
-    });
-    // Nobody supports current viewer, nor is anybody subscribed
-    if (platformWeights.length === 0) {
-      return localPlatform;
     }
 
-    const winningWeight = platformWeights[0].weight;
+    // Get weights for all of the platforms
+    return Promise.all(this.getAvailablePlatforms().map(platform => {
+      return this.calculatePlatformWeight_(platform, optionalFactor)
+          .then(weight => {
+            if (platform.getServiceId() === 'local') {
+              localWeight = weight;
+            }
+            return {
+              platform,
+              weight,
+            };
+          });
+    }))
+        .then(platformWeights => {
+          platformWeights.sort(function(platform1, platform2) {
+            return platform2.weight - platform1.weight;
+          });
+
+          const winningWeight = platformWeights[0].weight;
+          // Highest wins, local wins a tie
+          if (winningWeight > localWeight) {
+            return platformWeights[0].platform;
+          }
+          return localPlatform;
+        });
+  }
 
-    if (winningWeight > localWeight) {
-      return platformWeights[0].platform;
+  /**
+   * Calculate platform weight
+   * @param {!./subscription-platform.SubscriptionPlatform} platform
+   * @param {string=} optionalFactor if specified only calculate this factor
+   * @return {!Promise<number>}
+   * @private
+   */
+  calculatePlatformWeight_(platform, optionalFactor) {
```

Comment:
write test for this method

Reply:
this is covered by the tests on `selectApplicablePlatform_`